File: 38_string_arrange/main.py
# -*- coding: utf-8 -*-


def exchange(s, i, j):
    a = [e for e in s]
    a[i], a[j] = a[j], a[i]
    return ''.join(a)


def permutation_core(s, idx):
    if idx == len(s):
        print(s)
    else:
        for i in range(idx, len(s)):
            s = exchange(s, i, idx)
            permutation_core(s, idx + 1)
            s = exchange(s, i, idx)


def permutation(s):
    if len(s) == 0:
        return
    permutation_core(s, 0)
            
            
if __name__ == '__main__':
    permutation("abc")

File: 38_string_arrange/go/main.go
package main

import (
	"fmt"
)

func Exchange(s string, i, j int) string {
	ss := []byte(s)
	ss[i], ss[j] = ss[j], ss[i]
	return string(ss)
}

func PermutationCore(str string, idx int) {
	if idx == len(str) {
		fmt.Println(str)
	} else {
		for i := idx; i < len(str); i++ {
			str = Exchange(str, i, idx)
			PermutationCore(str, idx + 1)
			str = Exchange(str, i, idx)
		}
	}
}

func Permutation(str string) {
	if str == "" {
		return
	}
	PermutationCore(str, 0)
}

func main() {
	Permutation("abc");
}
